File: packages/i18nify-js/src/modules/banking/__tests__/validateGSTIN.test.ts
import validateGSTIN from '../validateGSTIN';

describe('banking - validateGSTIN', () => {
  describe('valid GSTINs', () => {
    it('should accept a valid GSTIN with state code 27 (Maharashtra)', () => {
      expect(validateGSTIN('27ABCDE1234F1Z5')).toBe(true);
    });

    it('should accept a valid GSTIN with state code 29 (Karnataka)', () => {
      expect(validateGSTIN('29ABCDE1234F2Z6')).toBe(true);
    });

    it('should accept state code 97 (Other Territory)', () => {
      expect(validateGSTIN('97ABCDE1234F1Z5')).toBe(true);
    });

    it('should accept state code 99 (Centre)', () => {
      expect(validateGSTIN('99ABCDE1234F1Z5')).toBe(true);
    });

    it('should accept a lowercase GSTIN (case-insensitive)', () => {
      expect(validateGSTIN('27abcde1234f1z5')).toBe(true);
    });

    it('should accept a GSTIN with surrounding whitespace', () => {
      expect(validateGSTIN('  27ABCDE1234F1Z5  ')).toBe(true);
    });
  });

  describe('invalid state codes', () => {
    it('should reject state code 25 (not assigned)', () => {
      expect(validateGSTIN('25ABCDE1234F1Z5')).toBe(false);
    });

    it('should reject state code 28 (not assigned)', () => {
      expect(validateGSTIN('28ABCDE1234F1Z5')).toBe(false);
    });

    it('should reject state code 00', () => {
      expect(validateGSTIN('00ABCDE1234F1Z5')).toBe(false);
    });

    it('should reject state code 39 (out of range)', () => {
      expect(validateGSTIN('39ABCDE1234F1Z5')).toBe(false);
    });
  });

  describe('invalid format', () => {
    it("should reject a GSTIN whose 14th char is not 'Z'", () => {
      expect(validateGSTIN('27ABCDE1234F1A5')).toBe(false);
    });

    it('should reject a GSTIN that is too short', () => {
      expect(validateGSTIN('27ABCDE1234F1Z')).toBe(false);
    });

    it('should reject a GSTIN that is too long', () => {
      expect(validateGSTIN('27ABCDE1234F1Z55')).toBe(false);
    });

    it('should reject a GSTIN with an entity char of 0', () => {
      // entity slot must be [1-9A-Z], 0 is not allowed.
      expect(validateGSTIN('27ABCDE1234F0Z5')).toBe(false);
    });

    it('should reject letters where the PAN numeric block is expected', () => {
      expect(validateGSTIN('27ABCDEABCDF1Z5')).toBe(false);
    });
  });

  describe('invalid arguments', () => {
    it('should throw for an empty string', () => {
      expect(() => validateGSTIN('')).toThrow(/GSTIN/);
    });

    it('should throw for a non-string input', () => {
      // @ts-expect-error testing runtime guard against non-string input
      expect(() => validateGSTIN(undefined)).toThrow(/GSTIN/);
    });
  });
});

File: packages/i18nify-go/modules/banking/validate_gstin_test.go
package banking

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

// ── ValidateGSTIN ──────────────────────────────────────────────────────────────

func TestValidateGSTIN(t *testing.T) {
	tests := []struct {
		name    string
		gstin   string
		want    bool
		wantErr bool
	}{
		{name: "valid GSTIN state 27 (Maharashtra)", gstin: "27ABCDE1234F1Z5", want: true},
		{name: "valid GSTIN state 29 (Karnataka)", gstin: "29ABCDE1234F2Z6", want: true},
		{name: "valid state 97 (Other Territory)", gstin: "97ABCDE1234F1Z5", want: true},
		{name: "valid state 99 (Centre)", gstin: "99ABCDE1234F1Z5", want: true},
		{name: "lowercase GSTIN is accepted", gstin: "27abcde1234f1z5", want: true},
		{name: "surrounding whitespace is trimmed", gstin: "  27ABCDE1234F1Z5  ", want: true},

		{name: "invalid state code 25", gstin: "25ABCDE1234F1Z5", want: false},
		{name: "invalid state code 28", gstin: "28ABCDE1234F1Z5", want: false},
		{name: "invalid state code 00", gstin: "00ABCDE1234F1Z5", want: false},
		{name: "out-of-range state code 39", gstin: "39ABCDE1234F1Z5", want: false},

		{name: "14th char not Z", gstin: "27ABCDE1234F1A5", want: false},
		{name: "too short", gstin: "27ABCDE1234F1Z", want: false},
		{name: "too long", gstin: "27ABCDE1234F1Z55", want: false},
		{name: "entity char 0 is rejected", gstin: "27ABCDE1234F0Z5", want: false},
		{name: "letters in PAN numeric block", gstin: "27ABCDEABCDF1Z5", want: false},

		{name: "empty string returns error", gstin: "", wantErr: true},
		{name: "whitespace-only returns error", gstin: "   ", wantErr: true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := ValidateGSTIN(tt.gstin)
			if tt.wantErr {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)
			assert.Equal(t, tt.want, got)
		})
	}
}
